File: codes/ReverseOnlyLetters917/solution.cs
public class Solution {
    public string ReverseOnlyLetters(string s) {
        // 双指针 easy
        int length = s.Length;
        char[] res = s.ToCharArray();
        int l = 0, r = length - 1;
        while(l < r){
            while(l < r && !IsAlpha(res[l])){
                ++ l;
            }
            while(r > l && !IsAlpha(res[r])){
                -- r;
            }
            char temp = res[l];
            res[l] = res[r];
            res[r] = temp;
            ++ l;
            -- r;
        }
        return new string(res);
    }

    public bool IsAlpha(char ch){
        return ch - 'a' < 26 && ch - 'a' >= 0 || ch - 'A' < 26 && ch - 'A' >= 0;
    }
}

File: codes/ReverseOnlyLetters917/Solution.java
package codes.ReverseOnlyLetters917;

public class Solution {
    //双指针滑动匹配
    public String reverseOnlyLetters(String s) {
        char[] template = s.toCharArray();
        int start = 0;
        int end = template.length-1;
        while(start < end){
            //每次使用数组前都需要判断是否越界
            while(start < end && !(template[start] >= 65 && template[start] <= 90 || template[start] >= 97 && template[start] <= 122)){
                start++;
            }
            while(start < end && !(template[end] >= 65 && template[end] <= 90 || template[end] >= 97 && template[end] <= 122)){
                end--;
            }
            //swap
            char tmp = template[start];
            template[start] = template[end];
            template[end] = tmp;
            start++;
            end--;
        }

        return String.valueOf(template);
    }

    public static void main(String[] args) {
        String s = "ab-cd";
        Solution solution = new Solution();
        System.out.println(solution.reverseOnlyLetters(s));
        
    }
}
